app/core/src/main/java/com/fsck/k9/notification/NotificationChannelUtils.java

```diff
@@ -0,0 +1,113 @@
+import android.app.NotificationChannelGroup;
+import android.app.NotificationManager;
+import android.content.Context;
+import android.os.Build;
+import android.support.annotation.RequiresApi;
+
+import com.fsck.k9.Account;
+import com.fsck.k9.DI;
+import com.fsck.k9.Preferences;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class NotificationChannelUtils {
+
+    public interface ChannelId {
+        String MESSAGES = "messages_channel";
+        String OTHER = "other_channel";
+    }
+
+    public static void updateChannels() {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            return;
+        }
+
+        Context context = DI.get(Context.class);
+        Preferences preferences = DI.get(Preferences.class);
+
+        NotificationManager notificationManager =
+                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
+        if (notificationManager != null) {
+            List<Account> accounts = preferences.getAccounts();
+            removeChannelsForNonExistingOrChangedAccounts(notificationManager, accounts);
+            addChannelsForAccounts(notificationManager, accounts);
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    private static void addChannelsForAccounts(NotificationManager notificationManager,
+                                               List<Account> accounts) {
+        for (Account account : accounts) {
+            final String groupId = account.getUuid();
+            NotificationChannelGroup channelGroup = new NotificationChannelGroup(groupId,
+                    account.getDisplayName());
+
+            NotificationChannel channelMessages = getChannelMessages(groupId);
+            NotificationChannel channelOther = getChannelOther(groupId);
+
+            notificationManager.createNotificationChannelGroup(channelGroup);
+            notificationManager.createNotificationChannel(channelMessages);
+            notificationManager.createNotificationChannel(channelOther);
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    private static void removeChannelsForNonExistingOrChangedAccounts(
+            NotificationManager notificationManager, List<Account> accounts) {
+        HashMap<String, Account> existingAccounts = new HashMap<>();
+        for (Account account : accounts) {
+            existingAccounts.put(account.getUuid(), account);
+        }
+
+        List<NotificationChannelGroup> groups = notificationManager.getNotificationChannelGroups();
+        for (NotificationChannelGroup group : groups) {
+            final String groupId = group.getId();
+
+            boolean shouldDelete = false;
+            if (!existingAccounts.containsKey(groupId)) {
+                shouldDelete = true;
+            } else if (!existingAccounts.get(groupId).getName().equals(group.getName().toString())) {
+                // There is no way to change group names. Deleting group, so it is re-generated.
```

Comment:
This will also delete the `NotificationChannel`s and their settings. We should keep the channels and attach them to the new `NotificationChannelGroup`.

Reply:
I think this is fine. Here's an official snippet from the Android SDK source code.

/**
     * Deletes the given notification channel.
     *
     * If you {@link #createNotificationChannel(NotificationChannel) create} a new channel with
     * this same id, the deleted channel will be un-deleted with all of the same settings it
     * had before it was deleted.
     */
    public void deleteNotificationChannel(String channelId)

Reply:
:+1: